lkmpg.tex

```diff
@@ -1204,6 +1204,29 @@ \section{System Calls}
 So, if we want to change the way a certain system call works, what we need to do is to write our own function to implement it (usually by adding a bit of our own code, and then calling the original function) and then change the pointer at \cpp|sys_call_table| to point to our function.
 Because we might be removed later and we don't want to leave the system in an unstable state, it's important for \cpp|cleanup_module| to restore the table to its original state.
 
+To change the pointer in \cpp|sys_call_table|, we need to consider the control register.
+A control register is a processor register that changes or controls the general behavior of the CPU.
+For x86 architecture, the cr0 register has various control flags that modify the basic operation of the processor.
+The WP flag in cr0 is write protection.
+Once the WP flag is set, the CPU may not write into the read-only sections.
+To modify \cpp|sys_call_table| we need to disable the WP flag.
+Since kernel version >= v.5.3, the \cpp|write_cr0| function cannot be used because of the sensitive cr0 bits pinned by the security issue, so we have to provide the custom asm to bypass it.
+However, \cpp|sys_call_table| symbol was unexported to prevent misuse.
+But there have few ways to get the symbol, memory seeking, and \cpp|kallsyms_lookup_name|, here we use \cpp|kallsyms_lookup_name|.
+Unfortunately, since kernel version >= 5.11, \cpp|kallsyms_lookup_name| is also unexported, it needs certain trick to get the address of \cpp|kallsyms_lookup_name|.
+If \cpp|CONFIG_KPROBES| is enabled, we use Kprobes to insert the breakpoint to get the corresponding function address.
+Kprobe is a debugging mechanism.
```

Comment:
Improve the writing. You should learn from https://www.kernel.org/doc/Documentation/kprobes.txt